ENABLE Setting DP Debug level at startup

bluezd opened the pull request:
This patch aims to support setting dp debug level at dp startup by
defining environment variable `DEBUG_LEVEL="ctrl error"` in addtiton to the API way.

bluezd commented:
Could anybody help reviewing it ?

gfsuse reviewed (commented):
levels==0 could mean clear previous debug setting
eg.)
2022-05-16T17:33:15.242|DEBU|AGT|dp.dpSendMsgEx: - msg={"ctrl_set_debug":{"categories":["ctrl"]}} size=42
2022-05-16T17:33:14|DEBU|cmd|dp_ctrl_handler: "ctrl_set_debug" done
2022-05-17T17:22:49|DEBU|cmd|dp_ctrl_handler: "ctrl_set_debug":{"categories": []}
2022-05-17T17:24:00|DEBU|cmd|dp_ctrl_handler: "ctrl_set_debug" done

gfsuse reviewed (commented):
I have downloaded, build and test your pull request. 
This change will not work, because set DP_DEBUG will only add -d debug_option to dp process, however enforcer consists of agent and dp, when agent starts and it connects to dp, after connection is established agent will send ctrl_set_debug msg to dp, this will overwrite whatever the env set for dp in yaml file.

bluezd commented:
> 

Hi, if you specify `CTRL_PATH_DEBUG` environment value that means you expect enable dp `ctrl` debug level explicitly so that will override other settings.

So for user want to  set dp debug level at startup explicitly they got to set `DP_DEBUG` than `CTRL_PATH_DEBUG`.

bluezd commented:
> 
Agent in `taskDPConnect` will set dp debug level as below: 
https://github.com/neuvector/neuvector/blob/main/agent/engine.go#L1808

I added `levels==0` check to avoid overwritten debug level even `gInfo.agentConfig.Debug` is not set. But as you mentioned it will lead to clear debug level not work, I will figure out a way to fix it. Thanks for your info.



I added `levels==0`

gfsuse commented:
Hi @bluezd, FYI, besides set debug level at the yaml env, we also have CLI to set debug setting for enforcer
eg.)
kubectl exec -ti neuvector-manager-pod-xxxxxx -n neuvector -- cli 
>show enforcer
Total enforcers: 2
+--------------+----------------------------------------------------------------------------------------------------------+------------------------+---------------------+----------------------+---------------+------------------+----------------------+
| id           | name                                                                                                     | host_name              | version             | joined_at            | cluster_ip    | connection_state | disconnected_at      |
+--------------+----------------------------------------------------------------------------------------------------------+------------------------+---------------------+----------------------+---------------+------------------+----------------------+
| 1721f7f88f9f | k8s_neuvector-enforcer-pod_neuvector-enforcer-pod-7m4qh_neuvector_781281e8-5ef6-4085-9b13-98c150c13274_0 | ubuntu1604-k818worker1 | interim/master.xxxx | 2022-05-18T16:42:38Z | 192.168.1.4   | connected        | 2022-05-18T16:43:52Z |
| 0723a010fd2d | k8s_neuvector-enforcer-pod_neuvector-enforcer-pod-sk6r2_neuvector_dc9d334d-b320-4116-8f84-980b03a76e7e_0 | ubuntu1604-k818worker2 | interim/master.xxxx | 2022-05-18T16:42:34Z | 192.168.2.155 | connected        |                      |
+--------------+----------------------------------------------------------------------------------------------------------+------------------------+---------------------+----------------------+---------------+------------------+----------------------+
>set enforcer 0723a010fd2d debug -c ctrl -c policy -c log

gfsuse commented:
pls consider following case:
1.
            - name: CTRL_PATH_DEBUG
              value: "1"
            - name: DEBUG_LEVEL
              value: "policy log"
2.
            - name: CTRL_PATH_DEBUG
              value: "0"
            - name: DEBUG_LEVEL
              value: "ctrl policy log"

bluezd commented:
> pls consider following case: 1. - name: CTRL_PATH_DEBUG value: "1" - name: DEBUG_LEVEL value: "policy log" 2. - name: CTRL_PATH_DEBUG value: "0" - name: DEBUG_LEVEL value: "ctrl policy log"

Thanks for your comments, I re-submitted the patch, could you please review it ? @gfsuse

gfsuse commented:
agree to merge
